core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingThreadPoolWDLParser.scala

```diff
@@ -0,0 +1,98 @@
+package org.broadinstitute.dsde.rawls.jobexec.wdlparsing
+
+import java.util.concurrent.{Callable, ExecutorService, Executors, TimeUnit}
+
+import com.github.benmanes.caffeine.cache.Caffeine
+import com.typesafe.config.ConfigFactory
+import com.typesafe.scalalogging.LazyLogging
+import languages.wdl.draft2.WdlDraft2LanguageFactory.httpResolver
+import org.broadinstitute.dsde.rawls.RawlsException
+import scalacache.{Cache, Entry, get, put}
+import scalacache.caffeine.CaffeineCache
+import wdl.draft2.model.{WdlNamespaceWithWorkflow, WdlWorkflow}
+import wdl.draft2.parser.WdlParser.SyntaxError
+
+import scala.collection.JavaConverters._
+import scala.util.hashing.MurmurHash3
+import scala.util.{Failure, Try}
+
+
+object CachingThreadPoolWDLParser extends WDLParsing with LazyLogging {
+
+  // TODO: conf should be injected, not read directly. At least this is an object so it happens once.
+  private val conf = ConfigFactory.parseResources("version.conf").withFallback(ConfigFactory.load())
+  private val wdlParsingConf = conf.getConfig("wdl-parsing")
+
+  private val cacheMaxSize = wdlParsingConf.getInt("cache-max-size")
+  private val threadPoolSize = wdlParsingConf.getInt("parser-thread-pool-size")
+  private val threadPoolTimeout = wdlParsingConf.getInt("parser-thread-pool-timeout-seconds")
+
+  // set up cache for WDL parsing
+  /* from scalacache doc: "Note: If you’re using an in-memory cache (e.g. Guava or Caffeine) then it makes sense
+     to use the synchronous mode. But if you’re communicating with a cache over a network (e.g. Redis, Memcached)
+     then this mode is not recommended. If the network goes down, your app could hang forever!"
+   */
+  import scalacache.modes.sync._
+  private val underlyingCaffeineCache = Caffeine.newBuilder()
+    .maximumSize(cacheMaxSize)
+    .build[String, Entry[Try[WdlWorkflow]]]
+  implicit val customisedCaffeineCache: Cache[Try[WdlWorkflow]] = CaffeineCache(underlyingCaffeineCache)
+
+  // set up thread pool
+  val executorService: ExecutorService = Executors.newFixedThreadPool(threadPoolSize)
```

Comment:
I'm wondering if `newCachedTreadPool` would be more appropriate then `newFixedTheadPool`.

See: https://stackoverflow.com/questions/949355/executors-newcachedthreadpool-versus-executors-newfixedthreadpool

Basically `newFixedTheadPool` will _always_ keep 8 threads ready, no matter what. `newFixedTheadPool` will spin up new threads as needed and terminate old ones. It's possible for >8 people to load an uncached MC at the same time (thinking about the workshop use case) and have to wait in line if limited to 8 threads. `newCachedTreadPool` would ensure that all requests are handled (and subsequently cached).

Reply:
totally understand your thought process! If the objective is simply to get this in an independent execution context, I agree - but my goal here was explicitly to put a bound on how many requests could be loading an uncached MC simultaneously - impose a bottleneck/queue to avoid contention issues. I'm open to hearing this bottleneck is unnecessary!

Reply:
Ah I see. However the underlying issue didn't seem to be related to concurrency, just total # of requests. So enforcing a concurrency limit of 8 wouldn't actually protect against the issue, and may have the side effect of increasing user wait time in high-concurrency situations.

Then again, a fixed thread pool might still be a good idea to protect Rawls against many concurrent requests.

I also wonder if `CallableParser` should recheck the cache before running, in case it had to wait in line and a previous thread already cached the result.

Reply:
I moved to `synchronize` per-WDL - and with `synchronize` in place, I think we don't need the extra cache lookup in `CallableParser`, and it's safer and more appropriate to use `newCachedTheadPool`.